### PR TITLE
Renew PACER cookies after PacerLoginException

### PR DESCRIPTION
**Pull request by albertisfu**
This is an initial proposal of a solution for #2160 to handle `PacerLoginException` errors on sentry.

Some findings that motivated this proposal are:

- Sentry is logging all the `PacerLoginException` even if they occur within a task retry, so for each failed retry, we get a log on the sentry for that error.

- The task that is raising the most `PacerLoginException` errors is `cl.corpus_importer.tasks.get_and_process_free_pdf` it has 15 retries set.
- I made a test in my dev environment that consists on: 
 -- Getting cookies from PACER and setting the expiration time on Redis to "Never".
 -- Trying to download a document 15, 30, and 90 minutes later, it worked.
 -- Then trying again 180 minutes later, the download failed. `PacerLoginException` was raised.

This might indicate that PACER is invalidating cookies from its side after some time.

Based on that, The problem that might be happening in prod is the following:

**When scraping PACER free opinions.**
- New PACER cookies are generated on the first executed task if they don't exist on Redis.
- After some time PACER invalidated these cookies before they expire on Redis.
- Following tasks try to use those cookies but they are invalid so `PacerLoginException` errors are raised.
- The task continues retrying e.g: `get_and_process_free_pdf` has 15 retries supposed to be retried every 5 minutes (on dev it retries every 3 minutes, not sure if this is due to a celery setting), so eventually a retry is executed with new cookies, so the task succeeded and the document is properly downloaded.

The proposed solution for this is (It's on this PR): 
- When a `PacerLoginException` occurs we should delete PACER cookies and cache fresh new ones before retrying.
- Current tasks and following tasks will use these fresh cookies until a new `PacerLoginException` happen on a task that will renew cookies again.

_A concern that I have here is when deleting the cookies from Redis. 
Is it possible to occur a race condition that deletes the cookies on different tasks? In my test downloading multiple documents at the same time, only the first task renews the cookies, and the following use them, but I'm not sure about this behavior in an environment working with multiple celery workers._

For now, I added this fix to the task that is producing the most `PacerLoginException` errors on sentry `get_and_process_free_pdf`, so we could check how it helps.

Hopefully, this fix would reduce a lot the number of `PacerLoginException` but it won't eliminate it totally since there will be a few of `PacerLoginException` within a retry where we'll renew cookies. I'll check if it's possible to avoid logging exceptions that happen within a retry on sentry to avoid these logs.

 
**For Recap Fetch API**
Something similar might be happening with the difference that we don't have credentials to generate new cookies. I found some Pacer fetch queues objects that failed on sentry and were not successfully downloaded, and others that succeeded.
So tasks that succeeded might be due to a following API request that renewed the user cookies. Then within a next retry new cookies are used. 
If there is not a new request, no cookies renew and the task fails.

Since we don't have credentials to get new cookies, I think a possible solution might be to reduce the cookies expiration time on Redis e.g: 30 minutes instead of 60 minutes to renew cookies sooner, I haven't applied yet this change yet I would like to hear your opinion.

Also, more ideas about this issue are welcomed!

### CODE REVIEW

**Review by mlissner — COMMENTED** [review]
A couple quick comments. I'll follow up with some replies to your questions.

**Comment by mlissner** [comment]
> This might indicate that PACER is invalidating cookies from its side after some time.

Yes. PACER logs you out after an hour or two. I haven't figured out how long, but it's not that long and it's annoying, so it makes sense to need to re-login fairly regularly.

> Is it possible to occur a race condition that deletes the cookies on different tasks? 

Yes. So your question is: If we have multiple workers all downloading things and PACER invalidates a cookie on their server, what happens? I'm guessing here, but imagine two workers, each processing tasks:

 - Worker 1: Attempts to download a PDF. All good.
 - Worker 2: Attempts to download a PDF. All good.
 - Worker 1: Attempts to download a PDF. LoginException! Starts getting new cookies.
 - Worker 2: Attempts to download a PDF. LoginException! Starts getting new cookies.
 - Worker 1: Finishes getting new cookies. Attempts to download a PDF. All good.
 - Worker 2: Finishes getting new cookies. Attempts to download a PDF. All good.

Yeah, that's not great, but it's the way things are now, so I think we can live with it for now unless you have a clever solution off hand. I couldn't think of anything better than a lock, which I didn't love.

> I'll check if it's possible to avoid logging exceptions that happen within a retry on sentry to avoid these logs.

Please do.

> I think a possible solution might be to reduce the cookies expiration time on Redis e.g: 30 minutes instead of 60 minutes to renew cookies sooner.

Yes, this is a good idea, but what I'd like to do is check the remaining TTL for a cached cookie when we get a fetch request. That way, if it's expiring in the next five minutes, say, we can refresh it *before* it expires. That should help a lot, I think.

**Comment by albertisfu** [comment]
@mlissner Thank you for your suggestions! I've applied them, instead of deleting cookies `delete_pacer_cookies_from_cache` and then getting new cookies with `get_or_cache_pacer_cookies` now we just use `refresh_pacer_cookies` that get fresh cookies and update them on Redis.

 I only applied `refresh_pacer_cookies ` method for `pacer_scraper` user because it's necessary to have credentials to re-login into PACER.

Also applied the suggestion of checking the cookies remaining TTL on Redis, if it's less or equal to 300 seconds (5 minutes) we get new cookies and update the key. This logic is on `get_or_cache_pacer_cookies` so it would help API users and pacer_scraper.

I did many tests and I was able to reproduce the `PacerLoginException` getting documents with `scrape_pacer_free_opinions` command, I could confirm that PACER is invalidating cookies quite often, that's why we were getting a huge amount of these exceptions.

Because the error was so common I decided to print a debug on a  Juriscraper testing [version](https://github.com/albertisfu/juriscraper/commit/233c434f11e52885013ebee1d6f94d0cb8be3fe9) to discard it's no a different message other than login. So I build a CL dev image with this Juriscraper version and logged the HTML.

So yes, I can confirm that the HTML was due to invalid cookies:
```
<html><head><link rel="shortcut icon"  href="/favicon.ico"/><title>CAND-ECF</title>
<script type="text/javascript">var default_base_path = "/"; </script><script type="text/javascript">if (top!=self) {top.location.replace(location.href);}</script><link rel="stylesheet" type="text/css" href="/css/default.css"><script type="text/javascript" src="/lib/core.js"></script></head><body><style type="text/css">body { overflow:auto;}</style><SCRIPT LANGUAGE="JavaScript">
document.cookie="FilingStatus=; path=/; expires=Thu, 01-Jan-1970 00:00:01 GMT";
document.cookie="PacerFirm=; path=/; domain=.uscourts.gov; expires=Thu, 01-Jan-1970 00:00:01 GMT";
document.cookie="KEY=; path=/; domain=.uscourts.gov; expires=Thu, 01-Jan-1970 00:00:01 GMT";
document.cookie="KEY=; path=/; expires=Thu, 01-Jan-1970 00:00:01 GMT";
document.cookie="PublicTerminal=; path=/; domain=.uscourts.gov; expires=Thu, 01-Jan-1970 00:00:01 GMT";

if ("https://pacer.login.uscourts.gov/csologin/login.jsf?pscCourtId=CANDC&appurl=https://ecf.cand.uscourts.gov/doc1/035122180317".length > 0) {
location.assign("https://pacer.login.uscourts.gov/csologin/login.jsf?pscCourtId=CANDC&appurl=https://ecf.cand.uscourts.gov/doc1/035122180317");
                                        }
</SCRIPT></div></body></html>
```

Strangely this happens quite often, here are the results of a test:

```
2022-08-01 19:03:37,881: Trying to get PDF: 035022180317
2022-08-01 19:03:37,900: PacerLoginException while getting free docs. Retrying.
2022-08-01 19:03:37,903: Refreshing cookies, new cookies: 4MHf5daKo158...
Task waiting for a new retry....

2022-08-01 19:03:56,633: Trying to get PDF: 035022179947, with cookies: 4MHf5daKo158...
2022-08-01 19:03:56,668: PacerLoginException while getting free docs. Retrying.
2022-08-01 19:03:56,671: Refreshing cookies, new cookies: AmOzUbEBTaOrzIro...
Task waiting for a new retry....

2022-08-01 19:06:37,351: New retry for: 035022180317, using cookies: AmOzUbEBTaOrzIro... Got PDF document successfully.
2022-08-01 19:06:56,182: New retry for: 035022179947, using cookies: AmOzUbEBTaOrzIro... Got PDF document successfully.

```
As you can see the cookie got by the 035022180317 tasks was invalid just a few seconds later.
The good news is that `refresh_pacer_cookies` eventually will get valid cookies and the document download will succeed.

About the other problem where we were getting within PacerLoginException errors on the sentry for each of these retries.  Is also solved.

I updated Sentry to the last version, but the problem persisted.

So doing a bit more research and testing, I found the problem might be related to this [issue](https://github.com/getsentry/sentry-python/issues/370).
But this issue is not solved yet.

So for now the solution is to avoid calling a task within another task. 

So I had to convert `download_pacer_pdf_by_rd` to a regular function instead of a task.

I checked and in fact, we don't call this function as a task in any place (no delay or apply_async), so it was straightforward to convert to a regular function.

I just move the exceptions it had to the functions where it's called in order to catch those kinds of exceptions and retry tasks if necessary.

So we this tweak, no more `PacerLoginException` within retries are sent to sentry.

The `PacerLoginException` only will be raised and sent to sentry when max retries are reached.

For `get_and_process_free_pdf task`:
Previously we had:

```
    except PacerLoginException as exc:
        msg = "PacerLoginException while getting free docs."
        if self.request.retries == self.max_retries:
            logger.warning(msg)
            self.request.chain = None
            return None
        # Refresh cookies before retrying
        refresh_pacer_cookies(
            "pacer_scraper",
            username=settings.PACER_USERNAME,
            password=settings.PACER_PASSWORD,
        )
        raise self.retry(exc=exc)
```
With this approach when max retries are reached it fails silently since just a warning is logged and the task returns None.


I thought it might be better to let the `PacerLoginException` be raised when max retries are reached so it will be logged into sentry.
So now we have:
   ```
 except PacerLoginException as exc:
        msg = "PacerLoginException while getting free docs."
        logger.info(f"{msg} Retrying.")
        # Refresh cookies before retrying
        refresh_pacer_cookies(
            "pacer_scraper",
            username=settings.PACER_USERNAME,
            password=settings.PACER_PASSWORD,
        )
        raise self.retry(exc=exc)
```
With this approach celery let the PacerLoginException be raised if max retries are reached.
Let me know if this seems good to you.

For the recap fetch tasks, we are using auto-retry for `PacerLoginException` so now with the change on `download_pacer_pdf_by_rd` no errors within retries will be sent to sentry. Only `PacerLoginException` will be raised and sent to sentry if max retries are reached.

Even though now we'll refresh cookies at the beginning of a recap fetch request.  If cookies TTL are close to expiring. I'm concerned that this might be not enough for API users since as you could see above seems that sometimes PACER is invalidating cookies so fast, hopefully, retries will help.

File: cl/lib/pacer_session.py [hunk 1]
```diff
@@ -79,3 +82,27 @@ def get_pacer_cookie_from_cache(user_pk: Union[str, int], r: Redis = None):
...
+    :param client_code: The PACER client code of the user
+    :return: Cookies for the PACER user
+    """
+
+    r = make_redis_interface("CACHE", decode_responses=False)
+    cookies = log_into_pacer(username, password, client_code)
+    cookie_expiration = 60 * 60
+
+    r.set(session_key % user_pk, pickle.dumps(cookies), ex=cookie_expiration)
+    return cookies
```
Comment (mlissner) [inline cl/lib/pacer_session.py:108]: This violates DRY. Can you factor it out along with the code above into a new function?

**Review by mlissner — CHANGES_REQUESTED** [review]
This looks good. One comment and a very small refactor:

> As you can see the cookie got by the 035022180317 tasks was invalid just a few seconds later.

Could this be due to distributed computers/databases? The PACER system, you'll recall is more like 200 systems which share some common parts. I could imagine that you're logging into one court (we used to use CA9, but I forget if we still do), and then you're trying to download from another. My theory is something like this happens:

 - You: Request cookie from CA9 by providing credentials
 - PACER Node (CA9): Generates a cookie for you, returns it to you and creates a task to send it to the central PACER auth database
 - You: Try to download something from a different node.
 - PACER Node (not CA9): Not auth'ed `PacerLogInException`!
 - PACER Central DB: Gets cookie from CA9 PACER Node. Saves it.

This is all just a guess, but I could see how if you're running a fast script and it's hitting more than one PACER node, perhaps their distributed auth system can't keep up? 

Thank you!

File: cl/lib/pacer_session.py (hunk 1 above)
Comment (albertisfu) [reply cl/lib/pacer_session.py]: Done, I've merge this function into `get_or_cache_pacer_cookies`

**Comment by albertisfu** [comment]
Thanks! I've merged the refresh method into get_or_cache_pacer_cookies, that's simpler!

I think your theory is correct!

I did some more tests to confirm it. In this test I removed the refresh cookies method when getting a PacerLoginException, this is what happened:

- Delete cookies from cache.
- Downloading 4 documents from `cand`
- Getting new cookies.

-Document 1 failed, retrying...
-Document 2 failed, retrying...
-Document 3 failed, retrying...
-Document 4 succeed

3 minutes later:
-Document 1 failed, retrying...
-Document 2 succeed
-Document 3 failed, retrying...

3 minutes later:
-Document 1 succeed
-Document 3 failed, retrying...

3 minutes later:
-Document 3 succeed...

So yes, seems that documents are being downloaded from different nodes and some cookies need more time to be spread to all machines. So eventually, if there are enough retries available the task will succeed. This is what is happening now in production. Even though we have a lot of PacerLoginException on sentry, documents are being downloaded properly due to retries.

But the refresh cookies method seems that will help too.

When using the refresh cookies method seems that we need fewer retries to download documents properly according to some tests e.g:

- Delete cookies from cache.
- Downloading 4 documents from `cand`
- Getting new cookies.

-Document 1 failed, refresh cookies, retrying...
-Document 2 succeed
-Document 3 succeed
-Document 4 succeed

3 minutes later:
-Document 1 succeed

This would mean that there are cookies that are more likely to spread faster across nodes.

**Comment by mlissner** [comment]
That cookie analysis is pretty weird. If all your downloads from from cand, I wouldn't expect multiple nodes to be an issue. Maaaybe they have more than one node per court (and CAND is a big one), but, hm, not sure. I'm surprised we don't see this error from a user perspective. 

Anyhow, the rest looks good, so merging. Thank you!

**Review by mlissner — APPROVED** [review]

**Pull request merged**